plasma_framework/contracts/src/transactions/PaymentTransactionModel.sol

```diff
@@ -37,40 +35,58 @@ library PaymentTransactionModel {
-        require(rlpOutputs.length <= MAX_OUTPUT_NUM, "Transaction outputs num exceeds limit");
+        return fromGeneric(GenericTransaction.decode(_tx));
+    }
 
-        uint txType = rlpTx[0].toUint();
-        require(txType > 0, "Transaction type must not be 0");
+    /**
+     * @notice Converts a GenericTransaction to a PaymentTransaction
+     * @param genericTx A GenericTransaction.Transaction struct
+     * @return A PaymentTransaction.Transaction struct
+     */
+    function fromGeneric(GenericTransaction.Transaction memory genericTx) internal pure returns (PaymentTransactionModel.Transaction memory) {
+        require(genericTx.inputs.length <= MAX_INPUT_NUM, "Transaction inputs num exceeds limit");
+        require(genericTx.outputs.length <= MAX_OUTPUT_NUM, "Transaction outputs num exceeds limit");
 
-        bytes32[] memory inputs = new bytes32[](rlpInputs.length);
-        for (uint i = 0; i < rlpInputs.length; i++) {
-            bytes32 input = rlpInputs[i].toBytes32();
+        bytes32[] memory inputs = new bytes32[](genericTx.inputs.length);
+        for (uint i = 0; i < genericTx.inputs.length; i++) {
+            bytes32 input = genericTx.inputs[i].toBytes32();
             // Disallow null inputs
             require(uint256(input) != 0, "Null input not allowed");
             inputs[i] = input;
         }
 
-        PaymentOutputModel.Output[] memory outputs = new PaymentOutputModel.Output[](rlpOutputs.length);
-        for (uint i = 0; i < rlpOutputs.length; i++) {
-            PaymentOutputModel.Output memory output = PaymentOutputModel.decode(rlpOutputs[i]);
-            require(output.outputType != 0, "Output type must not be 0");
-            require(output.amount != 0, "Output amount must not be 0");
-            outputs[i] = output;
+        GenericTransaction.Output[] memory outputs = new GenericTransaction.Output[](genericTx.outputs.length);
+        for (uint i = 0; i < genericTx.outputs.length; i++) {
+            outputs[i] = decodeOutput(genericTx.outputs[i]);
         }
 
-        bytes32 metaData = rlpTx[3].toBytes32();
+        bytes32 metaData = genericTx.txData.toBytes32();
 
-        return Transaction({txType: txType, inputs: inputs, outputs: outputs, metaData: metaData});
+        return Transaction({txType: genericTx.txType, inputs: inputs, outputs: outputs, metaData: metaData});
+    }
+
+    /**
+     * @notice Decodes a transaction output
+     * @param output The RLP encoded output
+     * @return The decoded GenericTransaction.Output struct
+     */
+    function decodeOutput(RLPReader.RLPItem memory output) internal pure returns (GenericTransaction.Output memory) {
+        RLPReader.RLPItem[] memory outputRlpList = output.toList();
+        require(outputRlpList.length == 4, "Output must have 4 items");
+        GenericTransaction.Output memory decodedOutput = GenericTransaction.decodeOutput(outputRlpList);
+        require(decodedOutput.amount != 0, "Output amount must not be 0");
```

Comment:
We check that the amount can not be zero here and that that output type can not be zero in `GenericTransaction.decodeOutput`. Is this intentional and does that mean that we plan to have specific transactions that can have zero amount?

Reply:
It's intentional. GenericTransaction should make as few assumptions as possible about the transaction.